nomenclature/validation.py

```diff
@@ -16,48 +16,46 @@ def is_subset(x, y):
     return set(to_list(x)).issubset([u or "" for u in to_list(y)])
 
 
-def validate(dsd, df):
+def validate(dsd, df, dimensions):
     """Validation of an IamDataFrame against codelists of a DataStructureDefinition"""
 
     if not isinstance(df, IamDataFrame):
         df = IamDataFrame(df)
 
     error = False
 
-    # combined validation of variables and units
-    invalid_vars, invalid_units = [], []
-    for variable, unit in df.unit_mapping.items():
-        if variable not in dsd.variable:
-            invalid_vars.append(variable)
-        else:
-            dsd_unit = dsd.variable[variable]["unit"]
-            # fast-pass for unique units in df and the DataStructureDefinition
-            if dsd_unit == unit:
-                continue
-            # full-fledged subset validation
-            if is_subset(unit, dsd_unit):
-                continue
-            invalid_units.append((variable, unit, dsd_unit))
-
-    if invalid_vars:
-        log_error("variables", invalid_vars)
-        error = True
-
-    if invalid_units:
-        lst = [f"{v} - expected: {e}, found: {u}" for v, u, e in invalid_units]
-        log_error("units", lst)
-        error = True
+    if "variable" in dimensions:
+        # combined validation of variables and units
+        invalid_vars, invalid_units = [], []
+        for variable, unit in df.unit_mapping.items():
+            if variable not in dsd.variable:
+                invalid_vars.append(variable)
+            else:
+                dsd_unit = dsd.variable[variable]["unit"]
+                # fast-pass for unique units in df and the DataStructureDefinition
+                if dsd_unit == unit:
+                    continue
+                # full-fledged subset validation
+                if is_subset(unit, dsd_unit):
+                    continue
+                invalid_units.append((variable, unit, dsd_unit))
+
+        if invalid_vars:
+            log_error("variable", invalid_vars)
```

Comment:
Not sure I like the design of just logging the invalid values and the referring to them in the error. Having the invalid values directly in an error would be better in my opinion. I understand that getting all invalid values in one single go is more efficient but pydantic does it by wrapping all the errrors raised during validation into a single validation error that contains a list of all errors.
Maybe that's overkill though, but seeing as we already have a function called `log_error` which in effect acts as a custom error it might not be.

Reply:
I agree that this pattern is not very elegant, but I'd keep it for now and then tackle this in a follow-up after release 0.1 - can you create an issue?